apis/v1beta1/metrics.go

```diff
@@ -0,0 +1,226 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	meterName = "crd-metrics"
+)
+
+// Metric labels
+
+const (
+	prefix               = "opentelemetry_collector_"
+	receiversMetricName  = prefix + "receivers"
+	exportersMetricName  = prefix + "exporters"
+	processorsMetricName = prefix + "processors"
+	extensionsMetricName = prefix + "extensions"
+	connectorsMetricName = prefix + "connectors"
+	modeMetricName       = prefix + "info"
+)
+
+// TODO: Refactor this logic, centralize it.
+type components struct {
+	receivers  []string
+	processors []string
+	exporters  []string
+	extensions []string
+	connectors []string
+}
+
+// Metrics hold all gauges for the different metrics related to the CRs
+// +kubebuilder:object:generate=false
+type Metrics struct {
+	modeCounter       metric.Int64UpDownCounter
+	receiversCounter  metric.Int64UpDownCounter
+	exporterCounter   metric.Int64UpDownCounter
+	processorCounter  metric.Int64UpDownCounter
+	extensionsCounter metric.Int64UpDownCounter
+	connectorsCounter metric.Int64UpDownCounter
+}
+
+// BootstrapMetrics configures the OpenTelemetry meter provider with the Prometheus exporter.
+func BootstrapMetrics() (metric.MeterProvider, error) {
+	exporter, err := prometheus.New(prometheus.WithRegisterer(metrics.Registry))
+	if err != nil {
+		return nil, err
+	}
+	return sdkmetric.NewMeterProvider(sdkmetric.WithReader(exporter)), err
+}
+
+func NewMetrics(prv metric.MeterProvider) (*Metrics, error) {
+	meter := prv.Meter(meterName)
+	modeCounter, err := meter.Int64UpDownCounter(modeMetricName)
+	if err != nil {
+		return nil, err
+	}
+	receiversCounter, err := meter.Int64UpDownCounter(receiversMetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	exporterCounter, err := meter.Int64UpDownCounter(exportersMetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	processorCounter, err := meter.Int64UpDownCounter(processorsMetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	extensionsCounter, err := meter.Int64UpDownCounter(extensionsMetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	connectorsCounter, err := meter.Int64UpDownCounter(connectorsMetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Metrics{
+		modeCounter:       modeCounter,
+		receiversCounter:  receiversCounter,
+		exporterCounter:   exporterCounter,
+		processorCounter:  processorCounter,
+		extensionsCounter: extensionsCounter,
+		connectorsCounter: connectorsCounter,
+	}, nil
+
+}
+
+// Init metrics from the first time the operator starts.
+func (m *Metrics) Init(ctx context.Context, cl client.Client) error {
```

Comment:
is there any reason why init is decoupled from the newMetrics?

Reply:
No reason, merged in the NewMetrics function